Sources/KSCrashRecordingCore/KSCrashMonitor.c

```diff
@@ -0,0 +1,288 @@
+//
+//  KSCrashMonitor.c
+//
+//  Created by Karl Stenerud on 2012-02-12.
+//
+//  Copyright (c) 2012 Karl Stenerud. All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall remain in place
+// in this source code.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+
+#include "KSCrashMonitor.h"
+#include "KSCrashMonitorHelper.h"
+#include "KSCrashMonitorContext.h"
+
+#include "KSDebug.h"
+#include "KSThread.h"
+#include "KSSystemCapabilities.h"
+
+#include <memory.h>
+#include <stdlib.h>
+
+//#define KSLogger_LocalLevel TRACE
+#include "KSLogger.h"
+
+typedef KSCrashMonitorAPI* (*GetMonitorAPIFunc)(void);
+
+typedef struct
+{
+    KSCrashMonitorAPI** functions; // Array of MonitorAPIs
+    size_t count;
+    size_t capacity;
+} MonitorList;
+
+#define INITIAL_MONITOR_CAPACITY 10
+
+// ============================================================================
+#pragma mark - Globals -
+// ============================================================================
+
+static MonitorList g_monitors = {};
```

Comment:
@naftaly, any idead here? Should I use atomics instead?

Reply:
As we now have calls of these API methods from global constructor functions, let's add syncing. What do we use in other C code for it?

Effectively, as there's no removal of monitors, we need sync for 2 operations:
1. The whole `kscm_addMonitor` function
2. The count getter: `g_monitors.count`

Reply:
Yes, but in case of a crash in the synchronized code here, we can get a deadlock, as this code should also handle SINGAL or Mach crash of KSCrash. So I'm thinking of atomic operations here, but it gets complicated with structs

Reply:
One of the options is to skip the crash if something updates monitors. E.g. if it's a lock we can add "if locked – exit". This kinda makes sense as if something crashes while KSCrash is being setup it's reasonable to ignore such crash.

Reply:
🤔